javalib/src/main/scala/java/util/LinkedList.scala

```diff
@@ -0,0 +1,385 @@
+    else removeNode(getNodeAt(index))
+  }
+
+  def peek(): E =
+    peekFirst()
+
+  def element(): E =
+    getFirst()
+
+  def poll(): E =
+    pollFirst()
+
+  def remove(): E =
+    removeFirst()
+
+  def offer(e: E): Boolean =
+    offerLast(e)
+
+  def offerFirst(e: E): Boolean = {
+    addFirst(e)
+    true
+  }
+
+  def offerLast(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  def peekFirst(): E =
+    head.value
+
+  def peekLast(): E =
+    if (last ne null) last.value
+    else peekFirst()
+
+  def pollFirst(): E =
+    removeFirst()
+
+  def pollLast(): E =
+    removeLast()
+
+  def push(e: E): Unit =
+    addFirst(e)
+
+  def pop(): E =
+    removeFirst()
+
+  private def _removeOccurrence(iter: Iterator[E], o: Any): Boolean = {
+    var changed = false
+    val e = o.asInstanceOf[E]
+    while (iter.hasNext() && !changed) {
+      if (iter.next() === e) {
+        iter.remove()
+        changed = true
+      }
+    }
+
+    changed
+  }
+
+  def removeFirstOccurrence(o: Any): Boolean =
+    _removeOccurrence(iterator(), o)
+
+  def removeLastOccurrence(o: Any): Boolean =
+    _removeOccurrence(descendingIterator(), o)
+
+  override def listIterator(index: Int): ListIterator[E] = {
+    checkIndexOnBounds(index)
+    new ListIterator[E] {
+
+      private var last = -1
+      private var i = index
+
+      private var currentNode: Node[E] =
+        if (index == size) null
+        else getNodeAt(index)
+
+      private var lastNode: Node[E] =
+        if (currentNode eq null) getNodeAt(size - 1)
+        else null
+
+      def hasNext(): Boolean =
+        (i < size)
+
+      def next(): E = {
+        if (i > size) throw new NoSuchElementException()
+        last = i
+        i += 1
+
+        if (currentNode eq null)
+          currentNode = lastNode
+
+        lastNode = currentNode
+        currentNode = currentNode.next
+
+        lastNode.value
+      }
+
+      def hasPrevious(): Boolean =
+        (i > 0)
+
+      def previous(): E = {
+        if (i < 0) throw new NoSuchElementException()
+        i -= 1
+        last = i
+
+        if (currentNode eq null)
+          currentNode = lastNode
+        else
+          currentNode = currentNode.prev
+
+        lastNode = currentNode
+
+        lastNode.value
+      }
+
+      def nextIndex(): Int = i
+
+      def previousIndex(): Int = i - 1
+
+      def remove(): Unit = {
+        checkThatHasLast()
+        removeNode(lastNode)
+        if (last < i)
+          i -= 1
+        last = -1
+      }
+
+      def set(e: E): Unit = {
+        checkThatHasLast()
+        lastNode.value = e
+      }
+
+      def add(e: E): Unit = {
+        if (currentNode eq null) {
+          if (lastNode.next eq null) {
+            addLast(e)
+            lastNode = getNodeAt(size - 1)
+          } else if (lastNode.prev eq null) {
+            addFirst(e)
+            lastNode = getNodeAt(0)
+          }
+        } else {
+          addNode(currentNode, e)
+        }
```

Comment:
unfortunately I have to discriminate the position where I come from (left, right), I do not want to put that logic into ```addNode``` method because it is not needed by the list itself, but only for this crazy kind of iterators, but I will think if there is a way to simplify it a bit...